BMC: support lazy goto model

smowton opened the pull request:
This introduces `goto_functions_providert`, which an interface that gets function bodies by name (perhaps converting them on demand), then gives a const view of the function map created so far, intended for reporting. It enables us to hide from `bmct` whether we're using plain old `goto_functionst` or `lazy_goto_modelt`.

Alternative to https://github.com/diffblue/cbmc/pull/1904, which tried harder to avoid use of `goto_functionst`, with the risk that users would store references to it or iterators pointing into it which could be invalidated upon converting more methods and so adding to the function map. However, test-gen would have required a **lot** of code to be changed, so it was necessary to expose a `const goto_functionst &` in the end regardless.

smowton committed:
Lazy goto model: implement abstract_goto_modelt

This enables it to be used interchangeably with goto_modelt, with the first user
being bmct (specificially the symex-and-solve process)

smowton committed:
Use abstract_goto_modelt instead of goto_functionst in bmc reporting phase

This allows lazy_goto_modelt to be passed in instead of a goto_modelt's functions
map as was used previously.

smowton commented:
@tautschnig @mgudemann @peterschrammel this is ready for review.

mgudemann commented on src/cbmc/bmc.cpp:
```diff
@@ -595,29 +609,40 @@ void bmct::setup_unwind()
+/// Perform core BMC, using goto_functions provider to supply GOTO function
+/// bodies (perhaps created on demand).
+/// \param opts: command-line options affecting BMC
+/// \param symbol_table: global symbol table. Calls to goto_functions_provider
```
does marking it `const` make sense then?

mgudemann reviewed (approved):
LGTM, but I am probably not the expert on that part of the code

mgudemann commented:
@smowton maybe I missed it, is there a TG pointer-bump?

smowton commented:
https://github.com/diffblue/test-gen/pull/1563

thk123 commented:
In terms of difference between this and #1904 - what would the advantage of this PR be (given it tries less hard to protect people from a shifting goto_functions?

smowton commented:
Less code churn, essentially. #1904 tried to hide goto_functionst completely, but test-gen's usage of it at this stage is too widespread for that to be practical, so it made rewriting well-behaved reporting functions like symex-coverage and the bmc-all-properties report phase seem rather pointless.

thk123 commented:
The other change still seems to have value since it means it is very easy to see people using the dangerous method (and can even be labelled as such) - lmk which one you'd rather I review

smowton commented:
Both! The other is a superset of this one though, so please start here and do the other as and when you get the chance. Hoping for @peterschrammel or @tautschnig to give an opinion about whether they want the broader change.

peterschrammel commented on src/goto-programs/lazy_goto_model.h:
```diff
@@ -102,6 +107,15 @@ class lazy_goto_modelt : public can_produce_functiont
+  /// Use with care; concurrent use of get_goto_function will have side-effects
+  /// on this map which may surprise users, including invalidating any iterators
+  /// they have stored.
+  const goto_functionst &get_goto_functions() const
```
I'd mark this is deprecated.

peterschrammel reviewed (commented):
Does it make sense that `symex_coveraget` operates on a lazy model? Shouldn't you freeze it before reporting?

smowton commented:
@peterschrammel I used to freeze it, but now `--paths` exists, which means this code is re-entered several times. We could also improve how `--paths` interacts with reporting passes, so the reporting code only runs once, but that's a bigger change, and probably one for @karkhaz.

smowton commented:
@peterschrammel also if you want that function deprecated, you may prefer #1904

tautschnig commented on src/goto-programs/lazy_goto_functions_map.h:
```diff
@@ -117,6 +117,17 @@ class lazy_goto_functions_mapt final
 
+  /// Returns the GOTO function body for function with name 'id', or throws if
+  /// that function has not been converted or does not exist.
+  /// Contrast `at(const irep_idt &)`, which converts functions on demand.
```
I don't quite understand: it seems to be calling this very function, so what exactly is the difference?

smowton replied:
Ah, yes this comment is outdated. Will fix.

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -125,6 +115,16 @@ class goto_symext
+    const get_goto_functiont &get_goto_function,
+    const statet &saved_state,
+    symex_target_equationt *const saved_equation,
+    symbol_tablet &new_symbol_table);
```
Why has this moved (it just changed in one parameter, if I got this right)?

smowton replied:
Yes, I only moved it because a function was inserted splitting the two overrides apart, when surely they belong together to help the reader.

tautschnig reviewed (dismissed):
My apologies for having been slow to get to this. I'd first like to understand: is the main aim of this exercise to provide access to `goto_functionst`, but also restrict it to the minimum required? If so, could it also be solved by just exposing `goto_functionst`, or is that not possible?

It just all reminds me very much of what fefe once described as "Klopapier Programmierung" (toilet-paper programming). I accept that wrappers can be useful for safety purposes, but they don't always improve readability of the code. (The concern usually is about performance, but I don't see that issue here.) Also I don't immediately see a much better/much different solution. In summary: I'm not really happy, but I don't know about a better solution. I'll look at #1904 later on today.

smowton commented:
@tautschnig the idea is to bundle together the convert-on-demand function and the get-all-functions-for-reporting (which gives a goto_functionst). I thought bundling them like this was cleaner than just passing `get_goto_functiont` and `const goto_functionst &` to `do_language_agnostic_bmc` and just commenting that one unexpectedly has side-effects on the other. This way it gets a single parameter whose properties are more obviously bound together. #1904 takes the idea further and keeps them bound together for the reporting phase, but has to give up at the last because test-gen (and perhaps other projects) demand too much of the `goto_functionst` interface.

I do however have one final idea, if you prefer: this was made tricky by the fact that `--paths` causes the reporting phase to run more than once, which @karkhaz noted was partially desirable because it means you can get partial results even if you kill cbmc early. Well, we could report **only the hit goals / violated assertions** for each `--paths` iteration without needing `goto_functionst`, then freeze the lazy model (which was the original intent, to freeze it and get an ordinary goto_modelt once the loading phase is over) before running end-of-day reporting.

smowton commented:
Re: exposing `goto_functionst` btw, do you mean just exposing a field (in which case fine, but I don't think there's much difference between a field and a getter except that the latter is more flexible to change later), or do you mean the passing `(get_goto_functiont, const goto_functionst &)` to `do_language_agnostic_bmc` scenario I described above?

tautschnig commented:
> Re: exposing `goto_functionst`

Indeed there isn't much of a difference between a field and a getter. Thinking about all this again I am starting to wonder about the growing number of interfaces that seemingly aren't quite sufficient. I see:
* `can_produce_functiont`
* `lazy_goto_modelt` is-a `can_produce_functiont`
* `goto_modelt`
* `goto_model_functiont` has-a `goto_functionst`

Now we'd also get a `goto_functions_providert` which seemingly is unrelated to `can_produce_functiont` or `goto_model_functiont`. Moreover, it seems to mix `goto_functionst` and `lazy_goto_modelt` instead of `goto_functionst`/`lazy_goto_functions_mapt` or `goto_modelt`/`lazy_goto_modelt`.

Where am I wrong?

tautschnig commented:
My apologies for the spam, now that I was able to sit down and review both #1914 (again) and #1904. In particular @kroening has put in considerable effort to move towards `goto_modelt` instead of passing `goto_functionst` and `symbol_tablet` separately. The work here seems to be partly thwarting this effort. I believe there should be some common interface implemented by both `goto_modelt` and `lazy_goto_modelt`. A candidate is `can_produce_functiont`, or some replacement thereof. Maybe the solution is simply for `goto_functions_providert` (with the API proposed in this PR) to become that interface (and dropping `can_produce_functiont`).

smowton committed:
Add abstract_goto_modelt

This provides an interface satisfiable by both goto_modelt and lazy_goto_modelt.
It exposes the symbol table and function map, but with the caution that get_goto_function
may produce functions on demand, and thereby have side-effects on either of them.

smowton committed:
goto_modelt: implement abstract_goto_modelt

smowton commented:
@tautschnig I've consolidated the interfaces as you recommended. Please re-review:

* `goto_modelt` and `lazy_goto_modelt` both inherit from `abstract_goto_modelt`, a common base class that provides const accessors to the symbol table and GOTO functions map as well as a non-const `get_goto_function` that can create functions on demand.
* `can_produce_functiont` is absorbed into `abstract_goto_modelt`
* `wrapper_goto_modelt` is used to implement the legacy interface that takes a `goto_functionst &`, with `outer_symbol_table` assumed to belong to the same model

There are a couple of remaining awkward spots, but I argue they are pre-existing and shouldn't be solved in this PR (but should be solved eventually):

* The new `--paths` option has awkwardly inverted the usual control-flow of the BMC phase, which ought to be partially reverted: parts of `bmct`'s state (especially the overall goal map and symex coverage data) have path-global scope while other parts such as the equation generated by symex are path-local, so we should divide it into a path-global and a path-local class. However this effort isn't currently a priority for us.
* The symbol table belonging to the goto-model is split away from its corresponding goto-function map and stored as `bmct` class state, while the function map is passed as a parameter. They should be kept together instead. However the restructure would break external users as it would change function and constructor parameters.

peterschrammel commented on src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -504,14 +504,14 @@ int jbmc_parse_optionst::doit()
   {
     configure_bmc = [](
-      bmct &bmc, const goto_modelt &goto_model) { // NOLINT (*)
+      bmct &bmc, const symbol_tablet &goto_model) { // NOLINT (*)
```
Type/parameter name mismatch: that doesn't look right

peterschrammel reviewed (commented):
This looks much cleaner now.

peterschrammel commented:
@smowton, this will require an update of the TG bump once there is general consensus to go ahead with this solution.

smowton commented:
Already has one, but will update & rebase to check it still works

smowton commented:
@peterschrammel fixed, and rebased test-gen bump (https://github.com/diffblue/test-gen/pull/1563), ready for review.
@tautschnig this is ready for re-review

tautschnig commented:
Will take a look now, but in the meantime a question:
> The new `--paths` option has awkwardly inverted the usual control-flow of the BMC phase, which ought to be partially reverted: parts of `bmct`'s state (especially the overall goal map and symex coverage data) have path-global scope while other parts such as the equation generated by symex are path-local, so we should divide it into a path-global and a path-local class. However this effort isn't currently a priority for us.

I don't understand what's been changed about _control-flow_ when the remainder of the comment  seems to be talking about the data representation. Would you mind elaborating, or indeed doing so in a new issue to be opened? (Surely this is not something you should have to take on as a priority, but you seem to have spotted some problem so the world should know about that.)

tautschnig commented on src/goto-programs/abstract_goto_model.h:
```diff
@@ -0,0 +1,49 @@
+#include "goto_functions.h"
+#include <util/symbol_table.h>
+
+class abstract_goto_modelt
```
This class/interface deserves a comment, although it might be the same as the file-level comment.

tautschnig commented on src/goto-programs/abstract_goto_model.h:
```diff
@@ -0,0 +1,49 @@
+
+class abstract_goto_modelt
+{
+public:
```
Add a virtual destructor.

tautschnig commented on src/goto-programs/goto_model.h:
```diff
@@ -66,6 +67,68 @@ class goto_modelt
   }
 
   void unload(const irep_idt &name) { goto_functions.unload(name); }
+
```
That closing brace looks misplaced.

tautschnig commented on src/goto-programs/goto_model.h:
```diff
@@ -15,12 +15,13 @@ Author: Daniel Kroening, kroening@kroening.com
+class goto_modelt : public abstract_goto_modelt
 {
 public:
   symbol_tablet symbol_table;
```
Direct access ought to become deprecated as there is an accessor function now.

tautschnig commented on src/goto-programs/goto_model.h:
```diff
@@ -66,6 +67,68 @@ class goto_modelt
+
+  bool can_produce_function(const irep_idt &id) const override
+  {
+    return goto_functions.function_map.count(id);
```
I still prefer `find(id) != end` over this I'm-lazy-and-don't-want-to-type-so-much idiom.

tautschnig reviewed (approved):
Looks good to me! (See below for minor comments.)

smowton commented:
@tautschnig the old control-flow was:

```
Symex -> BMC -> reporting
```

The new control-flow is:

```
for each path:
  Symex -> BMC -> reporting
```

However it ought to be:

```
Path-global init
for each path:
  Symex -> BMC -> incremental reporting
Path-global reporting
```

I've written this up here: https://github.com/diffblue/cbmc/issues/1945

smowton commented:
Applied some of those nitpicks, will merge once CI passes!
